Cumulus 1560 workflows (#1320)
nasa/cumulus: docs/workflows/input_output.md

A task's Lambda function can be configured to include a Cumulus Message Adapter library which constructs input/output messages and resolves task configurations.     The CMA can then be included in one of several ways:

### Lambda Layer

In order to make use of this configuration, a Lambda layer must be uploaded to your account.  Due to platform restrictions, Core cannot currently support sharable public layers, however you can support deploying the appropriate version from [the release page](https://github.com/nasa/cumulus-message-adapter/releases) via the AWS [Layers Interface](https://console.aws.amazon.com/lambda/home?region=us-east-1#/layers), *or* the provided CMA [Terraform](https://www.terraform.io/) module located at [tf-modules/cumulus-message-adapter](https://github.com/nasa/cumulus/tree/master/tf-modules/cumulus-message-adapter).

> Once you've deployed the layer, include the CMA layer in your  `cumulus_message_adapter_lambda_layer_arn` `tfvars` file if using the `cumulus` module for inclusion in all core Cumulus lambdas.

Author: Hmm, good point, need to remove the content on L30 too.   How about : ba0b4261a

Likewise, if including your own Lambda or ECS task Terraform modules, make sure to include that layer to include the CMA into your own code, and set a `CUMULUS_MESSAGE_ADAPTER_DIR` to `/opt` to make use of the CMA.

In the future if you wish to update/change the CMA version you will need to update the deployed CMA, and update the layer configuration for the impacted Lambdas as needed.}.

***Please Note***: Updating/removing a layer does not change a deployed Lambda, so to update the CMA you should deploy a new version of the CMA layer, update the associated Lambda configuration to reference the new CMA version, and re-deploy your Lambdas.

### Manual Addition

> You can include the CMA package in the Lambda code in the `cumulus-message-adapter` sub-directory in your lambda `.zip`, for any lambda runtime that includes a `python` [runtime](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html).  `python 2` is included in lambda runtimes that use Amazon Linux, however `Amazon Linux 2` will not support this directly.

Reviewer: ```suggestion
You can include the CMA package in the Lambda code in the `cumulus-message-adapter` sub-directory in your lambda `.zip`, for any Lambda runtime that includes a `python` [runtime](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html).  `python 2` is included in Lambda runtimes that use `Amazon Linux`, however `Amazon Linux 2` will not support this directly.
```

Also, should the first `python` in this message be `python 2`?

Author: No, we explicitly support 2.7, 3.6 for the CMA *for now*

As of Cumulus > 1.15 and CMA > v1.1.1, configuration of the CMA is expected to be driven by AWS Step Function Parameters.

> Using the CMA package with the lambda by by any of the above mentioned methods (Manual, Kes, Lambda Layers) requires configuration for it's various features via a specific Step Function Parameters configuration format (see sample workflows in the [examples cumulus-tf source](https://github.com/nasa/cumulus/tree/master/example/cumulus-tf) for more examples):

Reviewer: ```suggestion
Using the CMA package with the Lambda by any of the above mentioned methods (Lambda Layers, manual) requires configuration for its various features via a specific Step Function Parameters configuration format (see sample workflows in the [examples cumulus-tf source](https://github.com/nasa/cumulus/tree/master/example/cumulus-tf) for more examples):
```

> ```json

Reviewer: Can we update this to be a full JSON object to satisfy the Github markdown parser?

Author: 88034bcbb

In the workflow configuration (defined under the `task_config` key), each task has its own configuration, and it can use URL template as a value to achieve simplicity or for values only available at execution time. The Cumulus Message Adapter resolves the URL templates (relative to the event configuration key) and then passes message to next task. For example, given a task which has the following configuration:

```json

> Discovery:

Reviewer: Is this line extraneous?

{
  "Parameters": {
    "cma": {
      "event.$": "$",
      "task_config": {
        "provider": "{$.meta.provider}",
        "inlinestr": "prefix{meta.foo}suffix",
        "array": "{[$.meta.foo]}",
        "object": "{$.meta}"
      }
    }
  }
}
```

*and* and incoming message that contains:

```json

> "meta": {

Reviewer: This isn't a full JSON object in this snippet. Your call if it should be for consistency

Author: Fixed (plus some minor formatting nits): 4a465b818

Author: 9384f88a7

In the workflow configuration (defined under the `task_config` key), each task has its own configuration, and it can use URL template as a value to achieve simplicity or for values only available at execution time. The Cumulus Message Adapter resolves the URL templates (relative to the event configuration key) and then passes message to next task. For example, given a task which has the following configuration:

```json

> Discovery:

Author: ```suggestion
```

}
```

> ## Additional Feature

Author: One might argue this header is unneeded, open to suggestions to an alternate location for this info.

Reviewer: I think it's fine to leave it as `Additional features`, even if there is only one right now. I'm not sure where a better location would be

Author: ```suggestion
## Additional features
```
